### ISSUE TITLE
Feature Request: Fast Tool Gating & Single-Pass Selection via Prefill Logit Slicing

### ISSUE THREAD

**Issue by mattepiu** [ggml-org/llama.cpp#29022]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

## Motivation
In modern agentic workflows (e.g., ReAct, function calling, tool use), models spend significant wall-clock time autoregressively generating tokens just to make discrete routing decisions (e.g., spelling out `<tool_call>\n{"name": "get_weather"...}`).

## What is the proposed solution
A special agent mode to fasten tool calls through "Jev-like" hybrid mode

Currently, when tools are provided in `llama-server` or `llama-cli`:
1. The engine prefills the prompt.
2. It enters the autoregressive decode loop ($O(N)$ sequential token passes) using either GBNF grammar constraints or hoping the chat template emits tool tokens correctly.
3. If the model chooses *not* to call a tool, or chooses a simple tool, we still pay the latency penalty of sequential token decoding and grammar state-machine transitions before knowing the routing outcome.

Because the prompt prefill pass **already computes the full KV cache and logits for the final prompt token**, we can resolve binary tool-gating decisions and tool-name enums in **$O(1)$ single forward passes** via logit slicing, falling back seamlessly to standard text generation without wasting the prefilled KV cache.

---

## Proposed Architectural Flow

```
[Prompt + Registered Tools] 
          │
          ▼
   llama_decode() (Prefill)
          │
          ▼
Inspect logits of last token: P(Tool Call) vs P(Chat/Text)
          │
   ┌──────┴────────────────────────────────┐
   ▼                                       ▼
P(Chat) > Threshold               P(Tool) > Threshold
   │                                       │
   │ (Zero KV cache waste)                 ▼
   │ Direct stream to user        Slice logits over registered 
   │ (Standard Autoregressive)    tool names: P(tool_1...tool_N)
                                           │
                                  ┌────────┴─────────────────┐
                                  ▼                          ▼
                         Bounded Enums/Booleans     Unbounded String Args
                         (Parallel Logit Slices)    (Switch to GBNF / AR loop)
```

### 1. Zero-Cost Gating Check (Prefill Inspection)
Right after `llama_decode()` on the prompt batch, before initializing the autoregressive decode loop:
* Inspect the logits of the final sequence position (`llama_get_logits_ith(ctx, -1)`).
* Measure the probability mass of the model's tool delimiter token (e.g., `<tool_call>`, `[TOOL]`, or `{` depending on the active Jinja chat template) against normal text tokens.
* **If text generation is chosen:** Continue the decode loop immediately. The KV cache is already fully initialized; overhead is virtually $0$ ms.

### 2. $O(1)$ Discrete Tool Selection
If the model indicates a tool call:
* Instead of sequentially generating the tool name string (e.g., `"name": "get_weather_data"` across multiple decode steps), the engine can mask the logits over the first tokens of the user-registered tools:
  $$\text{Softmax}\left(\text{logits}[\text{id}(\text{tool}_1), \text{id}(\text{tool}_2), \dots]\right)$$
* Select the candidate with highest calibrated confidence in a single step.

### 3. Argument Resolution Split
* **Bounded arguments (Enums, Booleans):** Handled via single-pass logit slicing.
* **Unbounded arguments (Raw strings, free text):** Switch back to standard autoregressive generation, with GBNF constrained only to that specific string argument.

---

## Key Benefits
1. **Dramatic Latency Reduction in Agent Loops:** Eliminates 10–50 sequential autoregressive steps for routing decisions, reducing decision latency to the prefill time (~15–40 ms on modern hardware).
2. **Zero Wasted Computation on Fallback:** Because the prompt prefill already took place, falling back to standard conversational text streaming incurs zero recomputation or KV-cache invalidation.
3. **No Syntax Failures for Tool Routing:** Eliminates malformed tool name keys and invalid enum tool selections at the root.

---

## Alternatives Considered
* **Full GBNF Grammar Decoding:** Guarantees syntax correctness, but incurs full sequential decode latency ($O(N)$ forward passes) even for basic if/else decisions.
* **Draft Model Speculative Decoding:** Requires loading a second model into VRAM and doesn't solve the structural inefficiency of using autoregression for discrete enum selection.

### Motivation

Save a lot of inference and compute when performing tool calls

### Possible Implementation

1. **`llama-server` (`/v1/chat/completions`):**
   * Introduce an optional flag, e.g., `--speculative-tool-gating` or a request parameter `"tool_choice": "auto_fast"`.
   * If enabled, check the top-K logits of the prefill output against the model's configured tool-call trigger tokens before attaching the full GBNF grammar parser.
2. **C API (`llama.h`):**
   * Expose a helper utility or sampler stage:
     ```c
     // Evaluates probabilities exclusively across a masked subset of candidate tokens
     // on the current prefilled context position without advancing the decode sequence.
     LLAMA_API void llama_sample_token_distribution(
         struct llama_context * ctx,
         const llama_token    * candidate_tokens,
         size_t                 n_candidates,
         float                * out_probs
     );
     ```

### Edge Case: Multi-Token Tool Names & Shared Prefixes

To handle tools that share initial tokens (e.g., `get_weather` vs `get_stock`):

1. **Trie-Based Logit Slicing:** Construct a Token Trie of registered tools. Non-branching prefixes are auto-advanced, and logit slicing only triggers at branching nodes (reducing steps from full JSON generation to Trie depth $D \le 2$).
2. **Batched Sequence Evaluation:** Alternatively, fork the existing context KV-cache across candidate tool sequences in a single batched `llama_decode()` pass to score cumulative log-probabilities in $O(1)$ time. (seems MTP/DFlash friendly to me)
3. **Index Token Aliasing:** In `llama-server`, map tools to discrete index tokens `[A, B, C]` within the tool definition template to ensure candidate IDs are always strictly 1 token.

**Comment by lloyal-research** [comment]
How will you disambiguate between what to mask for tools with common-ish prefixes? Lexer-style lookahead?

**Comment by mattepiu** [comment]
> How will you disambiguate between what to mask for tools with common-ish prefixes? Lexer-style lookahead?

Added to the main object 3 possible paths. I believe number 2 could benefit from MTP/DFlash setups.

**Comment by shockme** [comment]
There is a [vLLM pull request](https://github.com/vllm-project/vllm/pull/57250) to add Jev-like functionality based on DiffusionGemma.

**Comment by kishida** [comment]
I've implemented the Jev-like functionality for any models. It works very well.
https://github.com/kishida/llama.cpp/blob/jev/docs/jev.md

is it fit for this issue?
I know offcause the code can not be merged since it may many point not to fit llama.cpp culture.

**Comment by taka-tuos** [comment]
@kishida , I’d like to run the benchmarks included in that model against the GGML-based JEV-like implementation I’m developing locally, but it looks like `bench_models.py` hasn't been added. Could you please check on that?

**Comment by mattepiu** [comment]
@kishida great work! that would deserve to be merged to main and allow to add systemone compatible api to use with tool like jegrep (zero index semantic search). Only thing missing in your bench is batched performances, which should raise latency a bit but allow for great throughtput (at least looking to how other project like laya.cpp behave)

**Comment by zqstdy** [comment]
Recently, I have been working on an AI gateway to assist in managing AI systems. There are multiple areas where decision-making issues need to be addressed. While browsing hf, I noticed that the implementation of jeV is very effective, but there is no locally deployable llamacpp model available yet. Furthermore, even if I were to deploy an additional model to increase local resource consumption, it would still be difficult to use given my limited configuration. I was wondering if I could directly use the existing model to achieve this effect. At this time, I stumbled upon your @mattepiu  issue and thought your solution was excellent. I implemented it using GLM-5.3-Flash and the results were quite good. Later, after seeing @Kishida's implementation, I realized that his implementation was more comprehensive. I borrowed some methods from glm. I wonder if  @kishida  is interested in opening a pull request. If so, I would like to see if we can combine our strengths and complement each other's functionalities

Here is my own simple test
```json

{
  "messages": [
    {
      "role": "user",
      "content": [
        {
          "type": "text",
          "text": "1+1=?\r\nA:1\r\nB:2\nAnswer strictly A or B."
        }
      ]
    }
  ],
  "max_tokens": 1,
  "temperature": 0,
  "stream": false,
  "chat_template_kwargs": { "enable_thinking": false },
  "logit_gate": {
    "candidates": [
      {
        "label": "A",
        "text": "A"
      },
      {
        "label": "B",
        "text": "B"
      }
    ],
    "mode": "gate_only",
    "options": {}
  }
}
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "role": "assistant",
        "content": ""
      }
    }
  ],
  "created": 1790139946,
  "model": "J:\\Code\\Models\\Qwen_Qwen3.5-9B-Q6_K.gguf",
  "system_fingerprint": "b11076-1dd2c5d26",
  "object": "chat.completion",
  "usage": {
    "completion_tokens": 0,
    "prompt_tokens": 31,
    "total_tokens": 31,
    "prompt_tokens_details": { "cached_tokens": 27 }
  },
  "id": "chatcmpl-nadP6lxHOQ0t8QOK5miIpCutZUwimJVD",
  "logit_gate": {
    "best": "B",
    "prob": 0.9918978214263916,
    "fired": true,
    "threshold": 0.5,
    "confidence": 0.9320691072352119,
    "temperature": 1.0,
    "distribution": [
      {
        "label": "A",
        "prob": 0.00810212828218937
      },
      {
        "label": "B",
        "prob": 0.9918978214263916
      }
    ]
  },
  "timings": {
    "cache_n": 27,
    "prompt_n": 4,
    "prompt_ms": 370.722,
    "prompt_per_token_ms": 92.6805,
    "prompt_per_second": 10.789756205458538,
    "predicted_n": 0,
    "predicted_ms": 0.0,
    "predicted_per_token_ms": 0.0,
    "predicted_per_second": 0.0
  }
}
```

The following is the data tested and organized by glm
## Block 1 — Statistics

### Head-to-head: single-pass gate vs generative verdict

Same prompts on both channels; gate channel = one prefill pass over the candidate logits, no tokens decoded; generative channel = temperature 0, verdict parsed from the output. 26 command-safety cases (21 with human labels + 5 ambiguous that only count toward agreement) and 14 memory-contradiction pairs per model. Idle GPU, one model per server run, single RTX 3060 12 GB (2026-09-23).

| Model | Decision type | Agreement | Gate accuracy | Generative accuracy | Latency (gate / generative) |
|---|---|---|---|---|---|
| Qwen3.5-9B-Q6_K (production launch config) | command review (26) | 92% | 95% | 100% | 238 ms / 1165 ms |
| Qwen3.5-9B-Q6_K | memory contradiction (14) | 100% | 100% | 100% | 271 ms / 2557 ms |
| Qwen3.5-9B-UD-Q4_K_XL | command review (26) | 96% | 95% | 100% | 228 ms / 950 ms |
| Qwen3.5-9B-UD-Q4_K_XL | memory contradiction (14) | 100% | 100% | 100% | 254 ms / 1775 ms |
| Qwen3.5-0.8B-Q8_0 | command review (26) | 73% | 86% | 52% | 65 ms / 307 ms |
| Qwen3.5-0.8B-Q8_0 | memory contradiction (14) | 57% | 64% | 93% | 64 ms / 410 ms |
| Ornith-1.5-9B-Q5_K_M (non-Qwen family) | command review (26) | 88% | 95% | 95% | 217 ms / 981 ms |
| Ornith-1.5-9B-Q5_K_M | memory contradiction (14) | n/a¹ | 100% | 0/14 parseable¹ | 255 ms / 3075 ms |

¹ The Ornith generative channel did not follow the requested verdict format on any of the 14 pairs, so agreement is not computable. The gate channel — which reads the first-token distribution and cannot fail to produce an answer — was 14/14.

Notes:

- Accuracy is against human labels on the 21 labeled cases; the 5 ambiguous commands (`npm install`, `git reset --hard`, `move config.json ../`, `docker system prune -a`, `wmic …`) carry no label and only count toward agreement.
- The single labeled disagreement on both 9B Qwen quants is the same case: `rm -rf ./build`, p = 0.34 / 0.40 → committed "low" at the raw 0.5 threshold. On the non-Qwen model the same command scores p = 0.56 (correct side) — a model property, not a mechanism flaw.
- A small model is the honest boundary: at 0.8B the gate still beats its own generative channel on commands (86% vs 52% — the generative channel flags `ls -la` as risky), but the probability mass squeezes into 0.4–0.8 and contradiction detection degrades (64%). Separation, not direction, is what small models lose — which is exactly what temperature scaling is for.
- Per-case raw logs for every run are archived; happy to share.

### Calibration (per model, shipped as a JSON file keyed by the model path)

Two-stage, run offline on a labeled seed set (22 commands + 14 memory pairs):

1. **Probability layer** — pool all seeds, fit one temperature on raw `label_logits` by minimizing NLL (grid search). A boundary hit means the seed set is too easy; the fit refuses extreme values and falls back to T = 1 — which is where Qwen3.5-9B actually landed (already honest, consistent with what the benchmark in this thread suggests for this family).
2. **Decision layer** — pick the smallest uncertainty band around the 0.5 threshold whose committed-region accuracy ≥ 98%. Inside the band the gate abstains and the request falls back to the full generative reviewer; outside it, the gate's verdict is final.

Q6_K fit (archived): band = 0.05 → command set 22 cases: 20 committed, 20/20 correct, 2 abstained; contradiction 14 committed, 14/14 correct.

The mid-probability cases are what the band is for: `rm -rf ./build` sits at p = 0.33–0.40 across Qwen 9B quants (0.56 on the non-Qwen model). An early mis-commit at a too-narrow band is literally what motivated the uncertainty-band design.

### Vision compatibility (mtmd / mmproj)

- Qwen_Qwen3.5-9B-Q6_K + mmproj-f16, archived server log (2026-09-22).
- Classification on an image: candidates {"logo", "no"} → p(logo) = 0.912, `gate_only`.
- `gate_then_chat` with an image: hit branch fired (p = 0.999); miss branch fell through and generated normally (80 tokens).
- Vision requests without the parameter are untouched — the gate only activates when the field is present.
- Cost: 768 ms request-to-decision including image encoding; prompt eval alone 641.5 ms / 273 tokens.

---

## Block 2 — Interface

A `logit_gate` field on the existing completion endpoints — no new endpoint, no new slot type. Evaluated once on the last prompt token's logits in the same slot-lifecycle branch as embeddings/rerank, before any decoding; nothing is generated.

Accepted on both `/v1/chat/completions` and `/v1/completions`:

```json
"logit_gate": {
  "candidates": [
    { "label": "yes", "text": "yes" },
    { "label": "no",  "ids": [9405] }
  ],
  "mode": "gate_only",
  "threshold": 0.5,
  "options": { "temperature": 1.5, "temperature_scaling": true, "return_logits": false }
}
```

- `candidates` (1–256, unique non-empty labels): each is a `label` plus exactly one of
  - `text` — must tokenize to exactly one token, or
  - `ids` — explicit token ids; probability mass is summed over them, so extra ids act as aliases (first-token scoring for multi-token words).
- `mode`: `gate_only` (default) always returns the distribution and stops; `gate_then_chat` stops with the decision only when the best probability ≥ `threshold` — otherwise the **same request continues as normal generation** with the prefilled KV cache reused, zero waste. This is the fall-through flow from this issue's description.
- `threshold`: default 0.5, clamped to [0, 1]; applies to `gate_then_chat` only.
- `options.temperature` (> 0): effective temperature = explicit request value → model GGUF metadata key `jev.temperature` → 1.0. `temperature_scaling: false` ignores it entirely. `return_logits: true` additionally returns raw per-label logits (un-tempered log-sum-exp over the label's ids) for fitting a temperature offline.

Response when the gate fires (empty `content`, `finish_reason: "stop"`, `usage.completion_tokens: 0`):

```json
"logit_gate": {
  "best": "yes",
  "prob": 0.944,
  "fired": true,
  "threshold": 0.5,
  "confidence": 0.689,
  "temperature": 1,
  "distribution": [ { "label": "yes", "prob": 0.944 }, { "label": "no", "prob": 0.056 } ],
  "label_logits": { "yes": …, "no": … }
}
```

(`confidence` = 1 − normalized entropy, 1.0 for a single candidate; `label_logits` only with `return_logits`. Values from a live probe: "Is fire hot?".)

**Fall-through, live:** candidates {yes, no} with `threshold: 0.99` on "What is 2+3? Answer with the number only." → no `logit_gate` field in the response, `content: "5"`, `completion_tokens: 4` — the request just continued as generation.

Validation — HTTP 400 with the offending field named, e.g.:

```
Field 'logit_gate': logit_gate candidate 'P0': text must tokenize to exactly 1 token (got 2); pass ids explicitly for first-token scoring
```

Also rejected: empty or duplicate labels, more than 256 candidates, out-of-range/empty `ids`, invalid `mode`, non-positive `temperature`. At runtime, non-finite logits are rejected explicitly instead of producing NaN probabilities and an empty "best".

**Comment by luancaldeira** [comment]
Thanks for posting the implementation details, especially the same-request fall-through behavior. One evaluation question: normalized-entropy confidence measures how concentrated the distribution is over the supplied candidates, so it could be high even when the correct option is missing. Are you planning to report coverage and accuracy at each threshold with an explicit out-of-set or abstain case, separately from syntax validity? I would also be interested in how the threshold is selected and checked on held-out cases.
